Ordering by fields coming form posts_meta table

naz opened the pull request:
refs https://github.com/TryGhost/Ghost/issues/11729

This is an attempt to add support for ordering by fields coming form `posts_meta` table. It introduces  couple new concepts:
1. A modification to ordering plugin - uses table names along with field names to form order query (e.g. `posts_meta.meta_description DESC`). This is needed to be able to specify ordering by different than model's native table.
2. Adds a "eager loading" plugin - it's main purpose is extension of the query with a join to relation which is used in ordering query.

@rishabhgrg & @allouis would love for you to have a look at the solution I've come up with, maybe you could spot something that is not obvious in the eager loading plugin or could come up with better naming/conventions for configurations that were introduced. In general would like a second pair of eyes from a perspective of "would I be able to use this plugin/ordering thing out of the box on similar relation in the future" :) Thanks!

naz committed:
Extended ordering to include table name in ordered field name

- The information about the table name is needed to avoid using `tableName` within pagination plugin and would give path to having other than original table ordering fields (e.g. order by posts_meta table fields)

naz committed:
Added test case to check ordering on posts_meta fields

- Should be removed or modified for final version. Treat as temporary case
- Fixed regression test utils for posts. The util was not up to date with what the actual endpoint was returning atm

naz committed:
Added support for eagerLoaded relations

- Allows to extend query builder object with joins to related tables
which could be used in ordering (possibly in filtering later).
- Bookshelf does not support ordering/filtering by probepries coming
from relations, that's why this kind of plugin and query expansion is needed

naz commented on core/server/models/plugins/pagination.js:
```diff
@@ -90,6 +90,27 @@ paginationUtils = {
+     * @param {Bookshelf.Model} model instance of Bookshelf model
+     * @param {string} propertyName property to be inspected and included in the relation
+     */
+    handleRelation: function handleRelation(model, propertyName) {
```
This is the place where ordering populates `eagerLoad` relations array on the model. This allows for `eagerLoading` plugin to pick it up later and create needed join.

naz commented on core/server/models/post.js:
```diff
@@ -68,6 +68,13 @@ Post = ghostBookshelf.Model.extend({
 
+    relationsMeta: {
+        posts_meta: {
+            targetTableName: 'posts_meta',
```
Key names `targetTableName` and `foreignKey` were borrowed from naming used in bookshelf models internally

naz committed:
Fixed unit test for pagination plugin

allouis reviewed (commented):
This looks good to me, I just had that one question about the regex!

naz commented on core/server/models/plugins/order.js:
```diff
@@ -31,11 +31,16 @@ const order = function order(Bookshelf) {
                 direction = match[2].toUpperCase();
 
-                if (orderAttributes.indexOf(field) === -1) {
+                const fieldRegexp = new RegExp(`${field}$`);
```
Yupp, the format of ordeing fields will be `{table_name}.{orderable_field_name}`. The regexp tries to match with the ending of the name. An alternative would be doing a match on the `field === orderAttribute.split('.')[1]`. This alternative method might be less performant I think, but there's rarely more than couple dozen records expected in these "orderAttributes", so maybe worth going with readability instead of "magic-regexp" :thinking: Wdyt? 

Side-note, both methods have a limitation that it not obvious and probably needs a `//NOTE:`. It's about not being able to use fields with the same name in "native" and "related" tables (for example if both `posts` and `posts_meta` tables had `title` field, it would be impossible to order by `title`, but that's expected I think given ordering is done of resource's fields which are usually unique). There is no usecase for this yet, but might avoid some confusion later on.

vikaspotluri123 replied:
Could you use  ```value.endsWith(`.${field}`)``` in the find fn instead of splitting the string?

naz replied:
Good point @vikaspotluri123 ! Why is my brain coming up with complex solutions first when there's something so much easier :sweat_smile:

allouis replied:
>  It's about not being able to use fields with the same name in "native" and "related" tables

This is what I was thinking, are you able to access the table name from the plugin, to get around this? I'm wondering if we would have issues with the order of the `orderAttributes` array?

e.g. `['parent.title', 'child.title']` and `['child.title', 'parent.title']` would give different orderings

Can you guarantee the order of that array? Should we consider `sort`ing it so that we can?

naz replied:
> e.g. ['parent.title', 'child.title'] and ['child.title', 'parent.title'] would give different orderings

Yupp, exactly the problem we _might_ face in the future. My thinking at the moment is that incoming order parameters will always be unique and "YAGNI" rule. I'd leave an extended comment about the limitations for now and move on.

> Can you guarantee the order of that array?

"Fun" fact - we can't with current implementation :grimacing: (raising an issue for this because it will probably bite us) What I have discovered is we keep a hash instead of an array in `options.order` attribute and than [iterate over hash keys](https://github.com/TryGhost/Ghost/blob/2cbba946632cb658d179861b4bb9fff697407441/core/server/models/plugins/pagination.js#L181-L186) when ordering is added to the query.  
For future reference, illustration of why ordering is not guaranteed with object keys even if defined in certain order ([source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys)):
```
// array-like object with random key ordering
const anObj = { 100: 'a', 2: 'b', 7: 'c' };
console.log(Object.keys(anObj)); // console: ['2', '7', '100']
```

> Should we consider sorting it so that we can?

We :100:  should once the above problem is solved and we have a clear usecase in our resources/models.

naz replied:
I've tried creating some test cases to confirm the case or ignored order query parameters order, but seems like because we use lodash's `_.forOwn` the order has been kept correctly. That said still think that keeping information about something that has to maintain it's order in a hash is not a best choice, arrays should be used if the order of properties is something we care about.

naz committed:
Fixed the regression sorting test

- Sorting workd differently in MySQL and SQLite when doint it on related tables. The main goal of the test is to prove the set is sorted by 'meta_description' field, so the rest of the set does not have to be consistent across dbs as long as this condition is satisfied

naz committed:
Refactor attribute matching to use `endsWith` instead of a RegExp

naz committed:
Fixed ordering match condition to catch 'count.posts' attribute

naz committed:
Added note about lack of support for child relations with same property names

naz commented:
Will merge this set of changes as it solves the usecase in the main issue. The limitations raised with ordering could be raised as separate issues in the future.
